File: solutions/cpp/collatz-conjecture/1/collatz_conjecture.cpp
#include "collatz_conjecture.h"
#include <stdexcept>

namespace collatz_conjecture 
{

unsigned int steps(int n) {
    if (n <= 0)
        throw std::domain_error("Must be nonzero");
    if (n == 1)
        return 0;
    else if (n % 2 == 0)
        return 1 + steps(n / 2);
    else
        return 1 + steps(n * 3 + 1);
}

}  // namespace collatz_conjecture

File: solutions/csharp/collatz-conjecture/1/CollatzConjecture.cs
using System;

public static class CollatzConjecture
{
    public static int Steps(int number)
    {
        int count = 0;
        Validate(number);
        while (number > 1)
        {
            if (number % 2 == 0)
            {
                number = number / 2;
            }
            else
            {
                number = number * 3 + 1;
            }
            count++;
        }
        return count;
    }

    static void Validate(int number)
    {
        if (number <= 0) 
        {
            throw new ArgumentOutOfRangeException();
        }
    }
}